### PR TITLE
bpo-37951: Lift subprocess's fork() restriction (GH-15544)

### PR DESCRIPTION
**Pull request by tiran**
Signed-off-by: Christian Heimes <christian@python.org>

https://bugs.python.org/issue37951

### CODE REVIEW

File: Doc/library/subprocess.rst [hunk 1]
```diff
@@ -483,6 +483,12 @@ functions.
       The *start_new_session* parameter can take the place of a previously
       common use of *preexec_fn* to call os.setsid() in the child.
 
+   .. versionchanged:: 3.8
+
+      The *preexec_fn* parameter is no longer supported in subinterpreters.
+      The new restriction may affect applications that are deployed in
+      mod_wsgi, uWSGI, and other environments.
```
Comment (vstinner) [inline Doc/library/subprocess.rst]: I'm not sure that it's useful to document affected applications.

**Review by vstinner — APPROVED** [review]
LGTM. Valid approach according to https://bugs.python.org/issue34651#msg325302

I just added a minor comment.

File: Doc/library/subprocess.rst (hunk 1 above)
Comment (tiran) [reply Doc/library/subprocess.rst]: The list of affected applications makes it easier for affected users to find the culprit. In my experience it's mostly embedded WSGI servers and embedded applications that use subinterpreters. mod_wsgi and uWSGI are the most common ones.

**Comment by tiran** [comment]
I'll wait for ACK from either @gpshead or @ericsnowcurrently before I push the fix.

File: Modules/_posixsubprocess.c [hunk 2]
```diff
@@ -583,8 +583,9 @@ subprocess_fork_exec(PyObject* self, PyObject *args)
             &restore_signals, &call_setsid, &preexec_fn))
         return NULL;
 
-    if (_PyInterpreterState_Get() != PyInterpreterState_Main()) {
-        PyErr_SetString(PyExc_RuntimeError, "fork not supported for subinterpreters");
+    if ((preexec_fn != Py_None) &&
+            (_PyInterpreterState_Get() != PyInterpreterState_Main())) {
+        PyErr_SetString(PyExc_RuntimeError, "preexec_fn not supported for subinterpreters");
```
Comment (gpshead) [inline Modules/_posixsubprocess.c]: s/for/within/

File: Doc/library/subprocess.rst (hunk 1 above)
Comment (gpshead) [reply Doc/library/subprocess.rst]: When you say "no longer supported", here and in whatsnew, mention that use of it raises a RuntimeError.

Comment (gpshead) [reply Doc/library/subprocess.rst]: FWIW I do agree that most _users_ probably are not aware that they're in a subinterpreter, so listing some common examples of such environments is not a bad thing.

Comment (tiran) [reply Doc/library/subprocess.rst]: @gpshead The subinterpreter and whatsnew entries now mention RuntimeError.

File: Modules/_posixsubprocess.c (hunk 2 above)
Comment (tiran) [reply Modules/_posixsubprocess.c]: Yes, that makes sense. ETOOLAZY

**Comment by tiran** [comment]
@gpshead Please review again

**Review by gpshead — APPROVED** [review]

**Comment by hroncok** [comment]
@tiran @ambv Any chance this will be in 3.8.0b4?

**Comment by ambv** [comment]
It will be in b4, yes.

**Pull request merged**